### PR TITLE
Support metadata properties outside wptrunner

### PR DESCRIPTION
**Pull request by jgraham**
For each CI system we run in there's a set of metadata properties
required to distinguish different configurations. For example on
GitHub when running infrastructure tests, that's the browser product
name, whereas in vendor CI it's typically different platform
configurations.

Until now the properties used when updating the metadata have been
stored in the browser product definition in wptrunner. But given they
vary per system that doesn't make much sense. Instead it should be
possible to pass in the properties relevant to the metadata being
updated.

To do that this PR adds the concept of a "metadata properties"
file. That's a JSON file of the form

{
 "property": ["name"],
 "dependent": {"name": ["other"]}
}

which specifies which properties can be used for update (this matches
the data format of the `update_properties` functions in wptrunner
products).

By default a file name `update_properties.json` in the root metadata
will be used to specify the properties. This makes sense since it puts
the properties together with the relevant metadata on the filesystem.

For now reading the properties from wptrunner is still supported,
because users need time to transition to the new format. But the
intent is to remove that later.

### CODE REVIEW

**Comment by WeizhongX** [comment]
Hi James, thanks for the change. Can you pls explain a little bit how "dependent" will be used?

**Comment by jgraham** [comment]
It's already used by the metadata update code, and this is explained in the docs change in the patch, but in short it's for something like "platform version". If you have a test that gives different results on e.g. linux, windows 7 and windows 11 (or whatever) you could just distinguish those on the basis of platform version, which might be "`20.04`", "`7`" and "`11`" (although actually the windows versions are more complex than that). But that's not ideal; it's much more informative to know that it first depends on platform and then _if_ the platform is Windows then it also depends on version. So the `dependent_properties` key is just to form a two level tree of properties, which says you can't use the second-level properties in an automatic metadata update before first splitting the results on the top-level property i.e. in this case you can't use `version` without also using `platform`.

File: tools/wptrunner/docs/expectation.rst [hunk 1]
```diff
@@ -305,6 +230,137 @@ and less than or equal to 200 total pixels different. For any
...
+
+wpt provides the tool ``wpt update-expectations`` command to generate
+expectation files from the results of a set of test runs. The basic
+syntax for this is::
+
+  ./wpt update-expectations [options] [logfile]...
+
+Each ``logfile`` is a wptreport log file from a previous run. These
+can be generated from wptrunner using the ``--log-wptreport`` option
+e.g. ``--log-wptreport=wptreport.json``.
```
Comment (gsnedders) [inline tools/wptrunner/docs/expectation.rst:248]: Do we want to make allowances for `--log-raw` etc., or just only document the `--log-wptreport` case?

Comment (jgraham) [reply tools/wptrunner/docs/expectation.rst:248]: Let's just document the `--log-wptreport` case; `--log-raw` adds quite a lot of unnecessary parsing overhead so it shouldn't be used in most cases.

Comment (gsnedders) [reply tools/wptrunner/docs/expectation.rst:248]: Does it make sense to throw an error when we have any other format? IIRC, some of the confusion historically has been providing the wrong format (like TBPL) appears to silently work?

Comment (jgraham) [reply tools/wptrunner/docs/expectation.rst:248]: Probably, but that's more or less unrelated to this PR.

**Review by gsnedders — APPROVED** [review]
The only real question I had here was with the documentation changes; the code changes all look fine to me.

**Pull request merged**